src/main/scala/com/evolutiongaming/crypto/Crypto.scala

```diff
@@ -1,149 +1,202 @@
+    * - AES/CTR/NoPadding (128 bit key) cipher with IV
+    */
+  private object AES_V2 {
+    private val CipherTransformation = "AES/CTR/NoPadding"
+
+    def decrypt(value: String, privateKey: String): String = {
+      val ivWithEncryptedData = Base64.decodeBase64(value)
+      val skeySpec = aesSKey128bitWithSha256(privateKey.getBytes(UTF_8))
+      val cipher = Cipher.getInstance(CipherTransformation)
+      val blockSize = cipher.getBlockSize
+      if (ivWithEncryptedData.length < blockSize) {
+        throw new IllegalArgumentException("invalid data size")
+      }
+      val (iv, encryptedData) = ivWithEncryptedData.splitAt(blockSize)
+      cipher.init(Cipher.DECRYPT_MODE, skeySpec, new IvParameterSpec(iv))
+      new String(cipher.doFinal(encryptedData), UTF_8)
+    }
   }
 
   /**
-    * Generates the SecretKeySpec, given the private key and the algorithm.
+    * Current AES mode - V3:
+    * - no restrictions on key size - SHA256 hash is used to obtain key entropy
+    * - AES/GCM/NoPadding (128 bit key) cipher is used to provide authenticated encryption
     */
-  private def secretKeyWithSha256(privateKey: String, algorithm: String) = {
-    val messageDigest = MessageDigest.getInstance("SHA-256")
-    messageDigest.update(privateKey.getBytes("utf-8"))
-    // max allowed length in bits / (8 bits to a byte)
-    // For AES we hardcode keylength to 128bit minimum to not depend on environment security policy settings:
-    // it may vary between 128 and 256 bits which can yield different encryption keys if we don't
-    val maxAllowedKeyLength = if (algorithm == "AES") AesKeyBytesMaxSize else Cipher.getMaxAllowedKeyLength(algorithm) / 8
-    val raw = messageDigest.digest().slice(0, maxAllowedKeyLength)
-    new SecretKeySpec(raw, algorithm)
+  private object AES_V3 {
+    /*
+    implementation based on
+    https://proandroiddev.com/security-best-practices-symmetric-encryption-with-aes-in-java-7616beaaade9
+     */
+
+    private val CipherTransformation = "AES/GCM/NoPadding"
+    /*
+    same auth tag length as in
+    https://proandroiddev.com/security-best-practices-symmetric-encryption-with-aes-in-java-7616beaaade9
+     */
+    private val AuthTagLengthBits = 128
+    /*
+    for GCM IV a 12 byte random byte-array is recommend by NIST
+    https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-38d.pdf
+     */
+    private val IVLengthBytes = 12
+
+    def encrypt(value: String, privateKey: String): String = {
+      val skeySpec = aesSKey128bitWithSha256(privateKey.getBytes(UTF_8))
+      val iv = new Array[Byte](IVLengthBytes)
+      secureRandom.nextBytes(iv)
+      val parameterSpec = new GCMParameterSpec(AuthTagLengthBits, iv)
+      val cipher = Cipher.getInstance(CipherTransformation)
+      cipher.init(Cipher.ENCRYPT_MODE, skeySpec, parameterSpec)
+      val encryptedValue = cipher.doFinal(value.getBytes(UTF_8))
+      Base64.encodeBase64String(iv ++ encryptedValue)
```

Comment:
I would suggest including IV size in base64-encoded output, for it to be more resilient and future-proof, as IV can be from 8 till 2^(64-1) bits (multiples of 8). Hard-coding IV size now, in future, will force to create new object (like, AES_V4) when different IV size will be needed.

Reply:
Would 1 byte be enough to encode IV size? I means we effectively limiting it to 255 max.

Reply:
Yes. I do not expect that larger IV would be beneficial as it still gets hashed into 128 bits and then one has to take into account hash collision probability for security.

Reply:
Done, please check.
I made it in a way that 12 bytes min check is not needed as the size is encoded as an unsigned byte with 0 meaning 12 bytes size - this give max IV size of 255 + 12 bytes

Reply:
I agree with @arixmkii, there's no need for "0 meaning 12", just use 0-255 and reject 0-11. It will still be fine. And it is more transparent.

Reply:
Agree. I've changed length encoding as you suggested, please check